### DISCUSSION TITLE
Decompiler question

### DISCUSSION THREAD

**Discussion by ThorvaldAagaard** [NationalSecurityAgency/ghidra#5440] [Q&A]
When I decompile an old application Ghidra generates source code like this

![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/b5937549-8117-419d-a580-b1e587e3e6e0)

Looking at a reference to this function it seems obvious that it is just a call to _fwprintf
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/e0690e6a-e56c-4dbd-bf5d-18db147ba472)

What is the proper way of telling Ghidra that there is no conflict, and it should just generate a call to ```_fwprintf```?

And by the way it looks like the parameters are wrong as %d means there should be a parameter more, and looking at the disassembling I think there are 3 parameters push on the stack
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/4f60e33b-a454-4bf1-847d-aec5ad753b19)

**Comment by quietust** [comment]
As I understand it, the Analysis process includes a step which tries to identify common library functions, and if it finds a function that matches more than one, it will apply one of the names and apply a "FID_conflict:" prefix to the name (and create a header comment which lists the other names, along with information on where it got that information from).

The proper way of telling Ghidra that there is no conflict is to simply *rename the function* and remove the bookmark pointing at it.

If you've marked the function with the `__cdecl` calling convention _and_ checked the "varargs" option in the function editor (which both appear to have been done in this case) and it's still not inferring the correct number of parameters at call sites, then I'm not sure what would be the proper way to fix it.
A starting point would be to make sure that all other functions called from within that function have proper calling conventions and parameter counts defined, since the decompiler might get confused otherwise.

**Comment by ThorvaldAagaard** [comment]
I tried renaming the function to  ```_fwprintf```, resulting in this
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/040df20c-b77f-4a14-ae30-395da779e0b1)
so it know the symbol (as I expected)

But what I really needed was to change all references to that function to ```_fwprintf```
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/77c0a341-64e5-4f59-b30e-d16ef62d171c)

So I located the first reference to the Conflict function and clicked "Edit function" getting this
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/248a242f-8e46-4010-9be6-d34c424fc041)

I tried to change the funtion name to _fwprintf, but that resulted in this error
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/fdaa01bc-0175-4ec3-84f7-83ac63f37422)

Override signature was my next try, but no help there
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/5ead3ca0-7a4c-416a-a602-ae2a823c38ce)

I learned a bit more but did not get closer to a solution, and there is still missing a parameter, that should replace the %d in the format.

**Reply by quietust** [reply]
What does the function's definition look like in the _disassembly_ view?

**Comment by ThorvaldAagaard** [comment]
This is what i see for the function that should not be there as the call made should be to _fwprintf
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/9e00764e-d5be-4e5a-9cac-1e19c31d8911)

And when I look at where it is called I see the following:
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/9983947f-85c0-44d3-9dd2-b59b89b3e482)
where I am missing a paramter, and the Disassembled view for this is 
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/8bceb024-2970-4fd7-abd6-1c6e9718b68a)

and this Listing
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/8c2968b9-2be8-4192-986c-29689944b0ad)

**Reply by quietust** [reply]
Just below the parameters and stack variables, you can see three symbols: `_fprintf` and `_fwprintf` in gold, and `FID_confict:_fwprintf` in blue. The third one is the **primary** symbol, and the others are secondary. Simply put the cursor on the label you want to use (e.g. by clicking on it), Edit it (by pressing the "L" key), then check the box to mark it as the Primary symbol.

You can also select the `FID_confict:_fwprintf` symbol and Delete it and that'll automatically pick a different symbol as the default, though it might pick the wrong one (in which case you can delete that one too).

As for the calling convention problem, I've got nothing.

**Reply by ThorvaldAagaard** [reply]
Thank you, that works fine, but I see some issues.

Assume that the developer In the original code used both  ```_fprintf``` and ```_fwprintf```, and as we have seen the decompiler can't figure out what was used and creates the conflict. 

I would assume that the two functions would have two different addresses, so perhaps there should be two different conflicts, but absolutely not sure it is the case.

But inspecting the code I can find examples like this

![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/266888ba-ea65-4d54-9699-2365bd8624f3)
where all looks fine after the change, and athis
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/76cabe76-deb2-4963-a969-42f6c6ec6150)
where I assume that it should be wide characters written.

The issue with the missing parameters to the functions is troublesome and I find these 2 statements almost side by side
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/70b86636-7f4a-4850-845f-c7138221b121)
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/4153ba5d-6476-48bc-ab95-50925147f983)
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/5651b936-29e2-498e-a13d-de470fb8c465)
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/650a119c-36ec-4d82-9264-a843a58058ce)

My other concern is that _fprintf is part of the standard library, but the decompiler have decompiled the code, so is there a way I can stop the decompiler from generating code for the library functions.

An example is that in KERNEL32.DLL, there is a function HeapAlloc, that is wrapped in a library function like this
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/87a9b451-f463-4c55-9c27-92fe83000f37)
That is called from this
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/68b55824-cd31-496b-86e2-77614ea56298)
and that function is in winheap.h
![image](https://github.com/NationalSecurityAgency/ghidra/assets/10560652/063d57a3-a52e-4f80-a8e8-2c52b91cb415)
